File: analyzers/python/confianalyzer_python/analyzer.py
"""Main analyzer — walks all Python files in a repo and produces a full IR document."""

from __future__ import annotations

import ast
import hashlib
import json
import logging
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from fnmatch import fnmatch

from confianalyzer_python import __version__
from confianalyzer_python.ast_walker import FileWalker
from confianalyzer_python.config import ConfiAnalyzerConfig
from confianalyzer_python.framework_plugins import get_active_plugins, get_plugins_by_name

logger = logging.getLogger("confianalyzer_python")

_EXCLUDE_DIRS = {
    "__pycache__", ".venv", "venv", ".tox", ".eggs",
    "dist", "build", "node_modules", ".git", ".mypy_cache",
    ".pytest_cache", "egg-info",
}


def _should_skip_dir(name: str) -> bool:
    if name in _EXCLUDE_DIRS:
        return True
    if name.endswith(".egg-info"):
        return True
    return False


def _is_test_file(path: str) -> bool:
    base = os.path.basename(path)
    return base.startswith("test_") or base.endswith("_test.py") or base == "conftest.py"


def _matches_any_pattern(rel_path: str, patterns: list[str]) -> bool:
    """Check if a relative path matches any of the glob patterns."""
    for pattern in patterns:
        if fnmatch(rel_path, pattern):
            return True
    return False


def _discover_files(
    repo_path: str,
    config: ConfiAnalyzerConfig | None = None,
) -> list[str]:
    """Discover all .py files excluding unwanted directories and test files."""
    has_include = config is not None and len(config.include) > 0
    has_exclude = config is not None and len(config.exclude) > 0

    result: list[str] = []
    for root, dirs, files in os.walk(repo_path):
        # Prune excluded directories in-place
        dirs[:] = [d for d in dirs if not _should_skip_dir(d)]
        for f in files:
            if f.endswith(".py") and not _is_test_file(f):
                full_path = os.path.join(root, f)

                # Apply config-based include/exclude
                if has_include or has_exclude:
                    rel_path = os.path.relpath(full_path, repo_path)
                    if has_include and not _matches_any_pattern(rel_path, config.include):  # type: ignore[union-attr]
                        continue
                    if has_exclude and _matches_any_pattern(rel_path, config.exclude):  # type: ignore[union-attr]
                        continue

                result.append(full_path)
    result.sort()
    return result


def _file_hash(content: bytes) -> str:
    return hashlib.sha256(content).hexdigest()


def analyze_repository(
    repo_path: str,
    repo_name: str,
    verbose: bool = False,
    incremental_path: str | None = None,
    config: ConfiAnalyzerConfig | None = None,
) -> dict:
    """Analyze a Python repository and return the IR document dict."""
    repo_path = os.path.abspath(repo_path)
    py_files = _discover_files(repo_path, config)
    logger.info("Discovered %d Python files in %s", len(py_files), repo_path)

    # Load previous IR for incremental analysis
    previous_hashes: dict[str, str] | None = None
    previous_files: dict[str, dict] | None = None

    if incremental_path:
        try:
            with open(incremental_path, "r", encoding="utf-8") as fp:
                prev_doc = json.load(fp)
            previous_hashes = {f["relativePath"]: f["hash"] for f in prev_doc["files"]}
            previous_files = {f["relativePath"]: f for f in prev_doc["files"]}
            logger.info(
                "Loaded previous IR with %d files for incremental analysis",
                len(prev_doc["files"]),
            )
        except Exception as exc:
            logger.warning("Could not load previous IR from %s: %s", incremental_path, exc)

    files_ir: list[dict] = []

    for file_path in py_files:
        try:
            raw = open(file_path, "rb").read()
            source = raw.decode("utf-8", errors="replace")
        except Exception as exc:
            logger.warning("Skipping %s: %s", file_path, exc)
            continue

        rel_path = os.path.relpath(file_path, repo_path)
        file_hash = _file_hash(raw)

        # Incremental: reuse previous entry if hash matches
        if (
            previous_hashes is not None
            and previous_files is not None
            and previous_hashes.get(rel_path) == file_hash
        ):
            logger.debug("Reusing cached analysis for %s", rel_path)
            files_ir.append(previous_files[rel_path])
            continue

        logger.debug("Analyzing %s", rel_path)

        try:
            walker = FileWalker(source, file_path, repo_path)
            data = walker.walk()
        except SyntaxError as exc:
            logger.warning("Syntax error in %s: %s", rel_path, exc)
            continue

        # Apply framework plugins
        if config and config.plugins:
            plugins = get_plugins_by_name(config.plugins, data["imports"])
        else:
            plugins = get_active_plugins(data["imports"])
        if plugins:
            tree = ast.parse(source, filename=file_path)
            _apply_plugins(tree, data, plugins)

        file_ir: dict[str, Any] = {
            "path": file_path,
            "relativePath": rel_path,
            "language": "python",
            "size": len(raw),
            "hash": file_hash,
            "functions": data["functions"],
            "calls": data["calls"],
            "imports": data["imports"],
            "exports": data["exports"],
            "classes": data["classes"],
        }
        files_ir.append(file_ir)

    ir_doc: dict[str, Any] = {
        "$schema": "confianalyzer-ir-v1",
        "version": "1.0.0",
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "analyzer": {
            "name": "confianalyzer-python",
            "version": __version__,
            "language": "python",
        },
        "repository": {
            "name": repo_name,
            "rootPath": repo_path,
        },
        "files": files_ir,
    }

    return ir_doc


def _apply_plugins(
    tree: ast.Module,
    data: dict,
    plugins: list,
) -> None:
    """Walk the AST again to apply framework-specific enrichments."""
    # Build a name->ir lookup for functions
    func_map: dict[str, dict] = {}
    for f in data["functions"]:
        func_map[f["name"]] = f

    # Build a location-keyed lookup for calls using a 4-tuple to avoid
    # collisions (e.g. `session.query(User).all()` has two Call nodes
    # starting at the same position but with different end positions).
    call_map: dict[tuple[int, int, int, int], dict] = {}
    for c in data["calls"]:
        loc = c["location"]
        key = (loc["startLine"], loc["startColumn"], loc["endLine"], loc["endColumn"])
        call_map[key] = c

    for node in ast.walk(tree):
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            ir = func_map.get(node.name)
            if ir is not None:
                for plugin in plugins:
                    plugin.enrich_function(ir, node)

        if isinstance(node, ast.Call):
            key = (
                getattr(node, "lineno", 0),
                getattr(node, "col_offset", 0),
                getattr(node, "end_lineno", 0) or getattr(node, "lineno", 0),
                getattr(node, "end_col_offset", 0) or getattr(node, "col_offset", 0),
            )
            ir = call_map.get(key)
            if ir is not None:
                for plugin in plugins:
                    plugin.enrich_call(ir, node)

File: analyzers/go/analyzer/analyzer.go
package analyzer

import (
	"crypto/sha256"
	"fmt"
	"go/parser"
	"go/token"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/confianalyzer/analyzer-go/analyzer/plugins"
	"github.com/confianalyzer/analyzer-go/irtypes"
)

const (
	AnalyzerName    = "confianalyzer-go"
	AnalyzerVersion = "0.1.0"
	IRSchema        = "confianalyzer-ir-v1"
)

// AnalyzeRepository analyzes all Go files in a repository and produces an IrDocument.
func AnalyzeRepository(repoPath, repoName string, verbose bool) (*irtypes.IrDocument, error) {
	absPath, err := filepath.Abs(repoPath)
	if err != nil {
		return nil, fmt.Errorf("resolving repo path: %w", err)
	}

	modulePath := detectModulePath(absPath)
	registry := plugins.NewRegistry()

	var files []irtypes.FileIR
	var walkErrors []string

	err = filepath.Walk(absPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			walkErrors = append(walkErrors, fmt.Sprintf("walk error at %s: %v", path, err))
			return nil
		}

		// Skip directories
		if info.IsDir() {
			base := filepath.Base(path)
			if base == "vendor" || base == "testdata" || base == ".git" || base == "node_modules" {
				return filepath.SkipDir
			}
			return nil
		}

		// Only .go files, skip test files
		if !strings.HasSuffix(path, ".go") || strings.HasSuffix(path, "_test.go") {
			return nil
		}

		if verbose {
			fmt.Fprintf(os.Stderr, "Analyzing: %s\n", path)
		}

		fileIR, err := analyzeFile(path, absPath, modulePath, registry)
		if err != nil {
			walkErrors = append(walkErrors, fmt.Sprintf("parse error at %s: %v", path, err))
			return nil
		}

		files = append(files, *fileIR)
		return nil
	})

	if err != nil {
		return nil, fmt.Errorf("walking repository: %w", err)
	}

	doc := &irtypes.IrDocument{
		Schema:      IRSchema,
		Version:     "1.0.0",
		GeneratedAt: time.Now().UTC().Format(time.RFC3339),
		Analyzer: irtypes.AnalyzerMeta{
			Name:     AnalyzerName,
			Version:  AnalyzerVersion,
			Language: "go",
		},
		Repository: irtypes.RepositoryMeta{
			Name:     repoName,
			RootPath: absPath,
		},
		Files: files,
	}

	if len(walkErrors) > 0 && verbose {
		for _, e := range walkErrors {
			fmt.Fprintf(os.Stderr, "WARNING: %s\n", e)
		}
	}

	return doc, nil
}

func analyzeFile(path, repoRoot, modulePath string, registry *plugins.Registry) (*irtypes.FileIR, error) {
	fset := token.NewFileSet()
	file, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
	if err != nil {
		return nil, err
	}

	hash, err := hashFile(path)
	if err != nil {
		return nil, fmt.Errorf("hashing file: %w", err)
	}

	info, err := os.Stat(path)
	if err != nil {
		return nil, fmt.Errorf("stat file: %w", err)
	}

	relPath, _ := filepath.Rel(repoRoot, path)

	result := WalkFile(fset, file, modulePath, registry)

	functions := result.Functions
	if functions == nil {
		functions = []irtypes.FunctionIR{}
	}
	calls := result.Calls
	if calls == nil {
		calls = []irtypes.CallIR{}
	}
	imports := result.Imports
	if imports == nil {
		imports = []irtypes.ImportIR{}
	}
	exports := result.Exports
	if exports == nil {
		exports = []irtypes.ExportIR{}
	}
	classes := result.Classes
	if classes == nil {
		classes = []irtypes.ClassIR{}
	}

	return &irtypes.FileIR{
		Path:         path,
		RelativePath: relPath,
		Language:     "go",
		Size:         info.Size(),
		Hash:         hash,
		Functions:    functions,
		Calls:        calls,
		Imports:      imports,
		Exports:      exports,
		Classes:      classes,
	}, nil
}

func hashFile(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return fmt.Sprintf("%x", h.Sum(nil)), nil
}

func detectModulePath(repoPath string) string {
	goModPath := filepath.Join(repoPath, "go.mod")
	data, err := os.ReadFile(goModPath)
	if err != nil {
		return ""
	}
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "module ") {
			return strings.TrimSpace(strings.TrimPrefix(line, "module"))
		}
	}
	return ""
}
